### DISCUSSION TITLE
Toward PyScript v5

### DISCUSSION THREAD

**Discussion by WebReflection** [pyscript/pyscript#2111] [Proposals]
#### Background

I know we use [CalVer](https://calver.org/) for versioning but, behind the scene, PyScript uses *semver* convention to land on **npm** in its "*canary*" or "*developer preview*" fashion.

Maybe not everyone payed enough attention to this tiny detail but we are at version **0.4.53** right now.

I've tried to stick with semver logic to date so that the *minor* number incremented only on breaking changes ... which ultimately never happened in recent months, all patches are features, as that's how it should be behind a `v0` project.

### Current State

Thanks to the team and everyone else involved within, or around, this project, we have reached some sort of maturity where most new things are either features or fixes and no drastic changes are usually required (until they are ... but we already cover tons of different use cases and we should be proud of that, imho).

We've also kept hearing, and learning from, the community that on every valuable occasion mentioned pain points and/or solutions to common cases.

This discussion is around all the learning that ourselves, or our users, gained over the last year(s) of development and I would like to propose the only sensible breaking (yet, not really) changes that could bump minor before it's too late (we should've learned from outdated docs, examples, social media press, articles that won't work anymore, and so on ...).

## The Future

I would like to summarize in here what I think would be an even more wonderful future for this project, listing point after point where I think everyone would win in the long term and commenting on each point too.

Everyone feedback would be more than appreciated and nothing is written in stones here, it's just a couple of points that keep bugging me daily and that are, imho, worth considering, or discussing.

- [ ] **Polyscript no more** - as much as I love *polyscript* as a project, I inevitably ended up branching its logic to accommodate its main, if not only, targeting project: *PyScript*. The architecture behind is able to host any PL that targets WASM but, at the same time, various features are allowed/enabled only for *PyScript* use cases and features. I love the fact this is all transparent for our users, but the maintenance effort keeps increasing to avoid breaking other PLs that can't compete with the maturity that both *Pyodide* and *MicroPython* on WASM give us these days. If we kill all other random WASM runtimes that gained, to date, very little interest in being adopted by *Polyscript* itself, we can also remove that name from our docs and write better code that is ad-hoc for our *Python* focused solution: no strings attached.
- [x] **top-level await** - both *Pyodide* and *MicroPython* provide a `runPythonAsync` that allows top-level await, yet we execute everything behind `runPython` instead for no reason whatsoever. We provide a `<script type="py/mpy">` convention that easily simulate or replace `<script type="module">` in the standard Web world, yet we don't handle that **entry point** as something anyone could await on our exposed `async` APIs, such as *fetch*, recently *storage*, and whatnot. We also have a lot of logic involving *hooks*, where we need to disambiguate with `async` those that are meant to run in sync VS async code, yet they lead to repeated hooks for no real-world benefit to anyone. Here I am saying we should get rid of any sync VS async disambiguation and just always run **async** by default. I am not super sure we should offer a *sync* version of anything, as when you write code in your *HTML* nobody is waiting for that last line result, so we can just call it a day and drop all "brain-complexity" or disambiguation needed to write hooks. Hooks to date have been untouched by our users: too complex to reason about and **not portable** across projects, due the fact once a script has `async` attribute, they might even never be executed. Last, but not least, there are users confused by the `async` attribute as if it could ever break their `sync` expectations, so it's clear to me having always `async` code executed would be a win.
- [ ] **persistent SQLite3** - we just released `pyscript.storage` but for anything more advanced or heavy, that might not be the best we can do. I've recently learned the original `sqlite3` *WASM* target project uses Opaque FileSystem API to never need the whole thing in memory (RAM). It can actually run out of FS related operations through that opaque FS so it would play better on both mobile and more constrained (RAM speaking) devices. I think these two are not mutually exclusive, but I also think we should not interfere much with native SQLite potentials: we provide its API as it is and we grant persistent results on users' sessions after reloads or any other operation. I don't have in mind the whole API right now but I think we should provide a module, or helper, that does the right thing for our users without being just *virtual*, like it is for the FileSystem we're using instead.
- [ ] **persistent FileSystem** - because we have `pyscript.storage` already out, nothing really stops us from providing, through the same mechanism, a persistent FileSystem that will automatically inject on bootstrap those persistent files in the virtual FS, with no need to have these listed in a `config`. This would enable, to me, a closer 1:1 solution for anything that needs persistent data and yet prefer a FileSystem approach over any K/V based databases.

Hopefully I've exposed my main "*desires*" for the future of the project, and I would be more than happy to add any other idea to this list (or remove some).

**Comment by Neon22** [comment]
My specific requirements are heading down the PWA track. (Full application running in browser doing work locally-using workers for lengthy calcs).
I need:
 - it to be easier to write UIs (ltk is great but code heavy. Puepy is coming along well).
 - Support for the "File System Access API" so I can have users upload and download(sync) the work they are doing in the Browser to their local file system. (existing recipe is excellent but not multi-file)
   - https://wicg.github.io/file-system-access/
   - This may mean a limited set of compatible browsers for a time.
 - A way to make it easy to download the 'page' and run it locally using the browser and local storage. I.e. a PWA.
FWIW I very much appreciate the recent work on getting Workers to behave well, and the work on fetch

**Reply by mchilvers** [reply]
We're currently working on fully-offline PWAs in pyscript.com, so that should be available (as an opt-in) soon.
If it helps, pyscript works just fine in Electron apps too if you need full access to the local machine.

**Comment by mchilvers** [comment]
+1 on all your ideas @WebReflection!

**Comment by fpliger** [comment]
Thanks for this @WebReflection !  Below a few questions/considerations.

> Polyscript no more - as much as I love polyscript as a project, I inevitably ended up branching its logic to accommodate its main, if not only, targeting project: PyScript. The architecture behind is able to host any PL that targets WASM but, at the same time, various features are allowed/enabled only for PyScript use cases and features. I love the fact this is all transparent for our users, but the maintenance effort keeps increasing to avoid breaking other PLs that can't compete with the maturity that both Pyodide and MicroPython on WASM give us these days. If we kill all other random WASM runtimes that gained, to date, very little interest in being adopted by Polyscript itself, we can also remove that name from our docs and write better code that is ad-hoc for our Python focused solution: no strings attached.

Can you better elaborate? This proposal has a few layers/consequences that I'd like to make sure are very clear. There's the pure tech aspect (== remove polyscript, bring plugins layers (or something like that) to PyScript, etc...) but also the other "Product" side of it around what actual interpreters/PLs we want to support in the future, etc. 

Can you cover the different aspects of your proposal and what it concretely means?


>  top-level await - both Pyodide and MicroPython provide a runPythonAsync that allows top-level await, yet we execute everything behind runPython instead for no reason whatsoever. We provide a <script type="py/mpy"> convention that easily simulate or replace <script type="module"> in the standard Web world, yet we don't handle that entry point as something anyone could await on our exposed async APIs, such as fetch, recently storage, and whatnot. We also have a lot of logic involving hooks, where we need to disambiguate with async those that are meant to run in sync VS async code, yet they lead to repeated hooks for no real-world benefit to anyone. Here I am saying we should get rid of any sync VS async disambiguation and just always run async by default. I am not super sure we should offer a sync version of anything, as when you write code in your HTML nobody is waiting for that last line result, so we can just call it a day and drop all "brain-complexity" or disambiguation needed to write hooks. Hooks to date have been untouched by our users: too complex to reason about and not portable across projects, due the fact once a script has async attribute, they might even never be executed. Last, but not least, there are users confused by the async attribute as if it could ever break their sync expectations, so it's clear to me having always async code executed would be a win.

Am I understanding the proposal correctly that this would mainly affect the top level API around the `async` attribute but, from a Python user point of view, you basically won't be really limited by the change and just happen to run `sync` or `async` Python code however you like out of the box?

>  persistent SQLite3 - we just released pyscript.storage but for anything more advanced or heavy, that might not be the best we can do. I've recently learned the original sqlite3 WASM target project uses Opaque FileSystem API to never need the whole thing in memory (RAM). It can actually run out of FS related operations through that opaque FS so it would play better on both mobile and more constrained (RAM speaking) devices. I think these two are not mutually exclusive, but I also think we should not interfere much with native SQLite potentials: we provide its API as it is and we grant persistent results on users' sessions after reloads or any other operation. I don't have in mind the whole API right now but I think we should provide a module, or helper, that does the right thing for our users without being just virtual, like it is for the FileSystem we're using instead.

+1. Would that be a different type of `pyscript.storage` or you mean something different (and only to support SQLite3? Couldn't the same useful for duckdb for instance? (no idea really, never looked into it, just really asking))

> persistent FileSystem - because we have pyscript.storage already out, nothing really stops us from providing, through the same mechanism, a persistent FileSystem that will automatically inject on bootstrap those persistent files in the virtual FS, with no need to have these listed in a config. This would enable, to me, a closer 1:1 solution for anything that needs persistent data and yet prefer a FileSystem approach over any K/V based databases.

Yes please. Back to my question above, how much of this overlaps with the previous item?

**Reply by WebReflection** [reply]
if we want to support other PLs we also need these PLs to have a close enough feature parity with Pyodide and MicroPython.

Right now that's unfortunately not the case, so that branching also documentation, expectations and so on for other PLs so far not many had interest in using feels like a bit of a overhead we don't fully need.

The current state is that Polyscript is 100% needed and used only for 2 out of 4 interpreters it's offering and those interpreters are anyway incapable of offering everything we do or need, including events on the DOM and whatnot.

I am OK keeping things the way these are right now as polyscript has never been a real issue or bottleneck, but the fact I need to jump back and forward between two projects that are strictly coupled and don't benefit from all the extra things (Ruby or Lua in particular) and branches in other things (Pyodide and MicroPython) doesn't feel the most elegant way forward, it's instead branching a lot of logic (inevitably due different capabilities) and it might be easier to contribute or add features if only focused on PyScript needs. Again, this is the least of the problems and the least of the focus / interest I have ... all these ideas were for the long-term future, nothing is needed now.

> from a Python user point of view, you basically won't be really limited by the change and just happen to run sync or async Python code however you like out of the box?

correct. `async` now is just an extra attribute to add that's needed after errors due `await` in the wild ... it makes portability of ad-hoc modules cumbersome too because one needs to know if that file or module requires top-level await or not.

For PyScript, having that transparently available, specially on the Web where many JS modules or native APIs are already *async*, feels like the best default we can offer.

`runPython` is something we don't even offer out of the box, it's a great primitive to get a result out of python code when you manually evaluate it, but nowhere in our current offer we directly export that feature.

Dare I say, *Ruby* can only run async too, and so is for *PHP* ... if many interpreters landing on WASM are async only, we'd better just drop some sync related code and embrace that way, slightly simplifying Polyscript code too (or better, removing some code that would throw otherwise in other PLs).

If we need, internally, to still access those sync APIs, we can always do that with ease.

>  Would that be a different type of pyscript.storage or you mean something different

it would be more like `from pyscript import sqlite3` to me, in a way that works in both Pyodide and MicroPython without needing any foreign package that won't work across these two different runtimes.

We asked Damien to provide some build that includes sqlite3 and Pyodide has that already but AFAIK (and IIRC) both are volatile, erasing somehow the beauty and usefulness of SQLite 3 as db.

Arguably though, I already have persistent SQLite 3 as JS module, so that we could even lazily provide that and see how it goes, before re-implementing something more suitable, or easier to embed, or more Pythonic, for our stack.

Same goes for the Persistent FS ... about DuckDB I think it's based on SQLite anyway but for NoSQL solutions a key/value store like `pyscript.storage` already exists and I am not sure we really need anything more complex than that.

**Reply by WebReflection** [reply]
never mind, I finally got where the question around DuckDB came from ... well, here an example of how we could deliver it too: we write a nice to embed package and we host it in our CDN for PyScript users and we document it, example: https://pyscript.com/@agiammarchi/jolly-silence-copy/latest

**Reply by fpliger** [reply]
Ok, so...  In regards to the topics:

* **Polysript/Many PLs:** Let's chat during the next community call. 
* **async**: +1 on using `async` as default from me
* **storage/sqlite/duckdb**: I don't see these topics as one. I'd separate them as a topic and would start looking at them considering the current user needs and possible scenarios. The ultimate goal is provide users with options to solver their problems. Not all of them need to be backed into the platform, and having an example/recipe can be sufficient. For others with higher demand, we can consider more streamlined solutions baked into PyScript itself (or a module)

**Comment by Neon22** [comment]
If we had the "File System Access API", could we not host the sql db on user's host. Providing some persistence.

**Comment by WebReflection** [comment]
**Update** beside the fact top-level-await is done (since quite a while ago) I have investigated the sqlite3 case and ...

>  I've recently learned the original sqlite3 WASM target project uses Opaque FileSystem API to never need the whole thing in memory (RAM).

Accordingly with the current state of affairs, [OPFS](https://developer.mozilla.org/en-US/docs/Web/API/File_System_API/Origin_private_file_system) is entirely async, but the sqlite3 WASM port is everything but async.

For the time being, I've implemented a solution that uses the same `idb-map/sync` underlying layer/stack (the one we use for the storage) to persist sqlite databases too, simply because I've learned that:

  * The OPFS is subject to [browser storage quota restrictions](https://developer.mozilla.org/en-US/docs/Web/API/Storage_API/Storage_quotas_and_eviction_criteria)
  * Clearing storage data for the site deletes the OPFS
  * Browsers persist the contents of the OPFS to disk somewhere, but you cannot expect to find the created files matched one-to-one. The OPFS is not intended to be visible to the user.

Basically, going *OPFS* would benefit only one use case: the user doesn't have enough RAM to handle the whole DB in memory ... but because there is quota associated, I am not expecting such DB to be that huge or issues will raise anyway and, on top of that, not using the available RAM means nothing can be done synchronously.

In few words, if we are OK exploring this field synchronously for either main or workers use cases, I believe I have exactly what we need already to land an sqlite3 db that works consistently between Pyodide and MicroPython too, but if we really want to pursue the OPFS way, the whole API is going to be inevitably async for every single operation, even if in workers such sync behavior could be abstracted away (Atomics) so that it doesn't have to be that way ... but that opens a variant of the same storage between main and workers, while the current implementation I have would work seamlessly well on both cases.
